### PR TITLE
Fix the type of optional[int] in dataclass

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
https://flyte-org.slack.com/archives/C037MMU74FN/p1660670745909239

```
@dataclass
class MyDataclass(DataClassJsonMixin):
    value: Optional[int] = None

my_dataclass = MyDataclass(value=1)
```

`my_dataclass.value` appears to be converted to float and becomes 1.0 instead of remaining int and 1 as I expected.

Explicitly convert the `value` to int as well.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2775

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/core/type_engine.py [hunk 1]
```diff
@@ -1309,6 +1311,11 @@ def convert_json_schema_to_python_class(schema: dict, schema_name) -> Type[datac
 def _get_element_type(element_property: typing.Dict[str, str]) -> Type[T]:
     element_type = element_property["type"]
     element_format = element_property["format"] if "format" in element_property else None
+
+    if type(element_type) == list:
```
Comment (eapolinario) [inline flytekit/core/type_engine.py:1315]: How does this work for `Union`?

Comment (pingsutw) [reply flytekit/core/type_engine.py:1315]: [marshmallow](https://github.com/marshmallow-code/marshmallow/issues/1191#issuecomment-480831796) doesn't support union type

**Review by wild-endeavor — APPROVED** [review]
